### ISSUE TITLE
float8_e4m3fn: float32 cast gives NaN on overflow where other stacks saturate to 448 — intended semantics?

### ISSUE THREAD

**Issue by dmitrii-f-t27** [jax-ml/ml_dtypes#400]
**Repro** (ml_dtypes 0.5.4, numpy 2.5.2, python 3.14):

```python
import numpy as np, ml_dtypes
src = np.array([1000.0, 448.0, 464.0, 465.0, -1000.0], dtype=np.float32)
print(src.astype(ml_dtypes.float8_e4m3fn))
# [nan 448 448 nan nan]
```

**Observation.** float32 → `float8_e4m3fn` casts produce **NaN on overflow**: values up to 464.0 round down to the max finite 448, values from 465.0 up go to NaN. Other stacks — e.g. the tt-metal / AMD convention — instead **saturate to the max finite 448 (0x7E)** on the same inputs.

OCP MX v1.0 leaves FP8 overflow behavior implementation-defined, so both semantics are legal — but a model quantized through one stack and executed on hardware following the other convention will disagree silently, exactly at overflow.

**Questions:**
1. Is NaN-on-overflow the intended, stable semantics for `float8_e4m3fn` casts in ml_dtypes (rather than an artifact)?
2. Do you have a recommendation on which convention downstream implementers should treat as the reference?

**Context:** we maintain an open conformance-vector suite for low-precision numeric formats (https://github.com/gHashTag/t27, `conformance/`); six of our packs are cross-checked against ml_dtypes 0.5.4, and this is the one divergence we hit. We currently document both behaviors (`docs/KAPPA_FP8_E4M3.md`). Happy to contribute test vectors covering the overflow boundary if that would be useful.

Filed at Peter Hawkins' suggestion (email thread).

**Comment by hawkinsp** [comment]
Just thinking about this quickly this looks like https://github.com/jax-ml/ml_dtypes/issues/147

i.e., there are both saturating and unsaturating casts, and you're seeing a NaN from an unsaturated cast.

Closing because I think this is working as intended.

---

AI response, for what it's worth:

This is the same distinction discussed in https://github.com/jax-ml/ml_dtypes/issues/147. Section 5 of the OCP 8-bit Floating Point Specification (OFP8 v1.0) defines both saturating and non-saturating conversion modes, and ml_dtypes (like NumPy, XLA ConvertElementType, and MLIR arith.truncf) implements non-saturating conversion for standard type casts.

1. Intended semantics: Yes, NaN-on-overflow is the intended, stable behavior for unclipped type casts (.astype(ml_dtypes.float8_e4m3fn)). Because 0x7E (448.0) has an even LSB (0) and a step size (ULP) of 32.0, the exact Round-to-Nearest-Even (RNE) midpoint before the 0x7F (NaN) slot is 448 + 16 = 464.0. Thus values in [448.0, 464.0] round to even (448.0), while values > 464.0 overflow to NaN. (In C++, ml_dtypes implements both via ConvertImpl<From, To, kSaturate, kTruncate>, defaulting to kSaturate = false for casts.)

2. Recommendation for conformance suites: Conformance suites should treat saturating (satfinite, clamp to ±448.0) and non-saturating (nosat, overflow > 464.0 to NaN) as two distinct conversion operations rather than a divergence between stacks. Standard type casts in compiler/array stacks use non-saturating mode so overflows are not silently masked, whereas ML quantization workflows explicitly clip before casting (jnp.clip(x, -448.0, 448.0).astype(ml_dtypes.float8_e4m3fn)), which lowers to hardware saturating conversion instructions on GPUs/TPUs.
